api/src/main/java/io/opencensus/stats/StatsRecord.java

```diff
@@ -0,0 +1,63 @@
+/*
+ * Copyright 2017, OpenCensus Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.opencensus.stats;
+
+import io.opencensus.stats.Measure.MeasureDouble;
+import io.opencensus.stats.Measure.MeasureLong;
+import io.opencensus.tags.TagContext;
+import javax.annotation.concurrent.NotThreadSafe;
+
+/** A map from {@link Measure}s to measured values to be recorded at the same time. */
+@NotThreadSafe
+public abstract class StatsRecord {
+
+  /**
+   * Associates the {@link MeasureDouble} with the given value. Subsequent updates to the same
+   * {@link MeasureDouble} will overwrite the previous value.
+   *
+   * @param measure the {@link MeasureDouble}
+   * @param value the value to be associated with {@code measure}
+   * @return this
+   */
+  public abstract StatsRecord put(MeasureDouble measure, double value);
+
+  /**
+   * Associates the {@link MeasureLong} with the given value. Subsequent updates to the same {@link
+   * MeasureLong} will overwrite the previous value.
+   *
+   * @param measure the {@link MeasureLong}
+   * @param value the value to be associated with {@code measure}
+   * @return this
+   */
+  public abstract StatsRecord put(MeasureLong measure, long value);
+
+  /**
+   * Records all of the measures at the same time, with the current {@link TagContext}.
+   *
+   * <p>This method records all of the stats in the {@code StatsRecord} every time it is called.
+   */
+  public abstract void record();
+
+  /**
+   * Records all of the measures at the same time, with an explicit {@link TagContext}.
+   *
+   * <p>This method records all of the stats in the {@code StatsRecord} every time it is called.
+   *
+   * @param tags the tags associated with the measurements.
+   */
```

Comment:
Too wordy imho just say record(context) or recordInContext(context)

Reply:
I added "WithExplicitTagContext" to the method name because @songy23 thought that having an overloaded method to create the StatsRecord was unclear (https://github.com/census-instrumentation/opencensus-java/pull/676#discussion_r143073529), but I wonder if the distinction between the use cases is clearer now that `StatsRecord.record` takes the optional TagContext.  If the TagContext can only be provided at the same time that the stats are recorded, it should be clear which tags will be used.

/cc @songy23

Reply:
Agreed. In this situation, overloading record() with an optional TagContext is less ambiguous (than having it at newRecord()). I'm fine with renaming this and just use record().

Reply:
Thanks for the feedback.  I made a PR: #728